### ISSUE TITLE
decode_ec_private_key result fails in create_object

### ISSUE THREAD

**Issue by fzonneveld** [danni/python-pkcs11#77]
Using Utimaco HSM to load ec private key in DER format.

```
k = ec.decode_ec_private_key(pk_der)
k[Attribute.LABEL] = 'EC key'
session.create_object(k)
```
Fails.
Used asn1crypto to set the EC_PARAMS and VALUE attributes. This succeeds.

Difference between binary values:
EC_PARAMS=b'a00706052b81040022' (wrong)
EC_PARAMS=b'06052b81040022'
VALUE=b'04306f0eabcebfe3d9cc45a6e412aa7c08ae0e2a0c297cbde44f3eb0344fbe8c2431d4aaf5505ec178843cdcc662a48b826d' (wrong)
VALUE=b'6f0eabcebfe3d9cc45a6e412aa7c08ae0e2a0c297cbde44f3eb0344fbe8c2431d4aaf5505ec178843cdcc662a48b826d'

Can see for EC_PARAMS extra ASN1 encoding tag and for VALUE also extra ASN1 encoding tag 0x04, which is an ASN1 OCTET STRING.

**Comment by danni** [comment]
It looks like something is failing decoding variant types from ASN.1. There's either an extra variant marker, or we've failed to remove one. Where did you get this key from? Does OpenSSL decode it correctly?

**Comment by fzonneveld** [comment]
I looked at the problem a little bit closer and updated the ticket, please
see the ticket for details.
What I used to generate the key was "`openssl ecparam -genkey -name
secp256r1 | openssl ec -outform der > ec.der"
I use also use asn1crypto to extract the domain parameters and the key.

Thanks Frank,

**Comment by danni** [comment]
python-pkcs11 doesn't interpret the values except in `decode_ec_private_key`, it passes them straight on. The `.so` / the HSM then interprets the ASN.1 encoded value.

I'm guessing this is some disagreement within the vendor module about whether or not to include the variant marker [1]. Does the key you've generated/code you've written work with SoftHSM?

[1] The spec here simply says `ANSI X9.62 private value d`, and doesn't mention DER encoding. So I'm not actually sure I'm right. I've never tried importing a private key into a real HSM because the ones I've used wouldn't allow it.

**Comment by fzonneveld** [comment]
ASN.1 is abstract and needs to be encoded into BER/DER/XER/PER or any encoding version to be able to export/import again. With the dump() method from asn1crypto DER encoded value for the structure is created.

I do agree is confusing, and also does not help that  ASN X9.62 is a payed spec. Openssl ec generates 'The DER option with a private key uses an ASN.1 DER encoded SEC1 private key.' This references RFC5915.txt  https://www.ietf.org/rfc/rfc5915.txt. In this specifies the ECPriavteKey as an OCTET string, see b) in the introduction. OCTET string is  0x04 in DER encoding for ASN1, see http://luca.ntop.org/Teaching/Appunti/asn1.html.

I will try to get the SoftHSMv2 (https://github.com/opendnssec/SoftHSMv2) up and running to see if an EC key can be imported in there, in this way you can reproduce the error...

**Comment by danni** [comment]
Yep I understand that :) What I meant is that the PKCS#11 spec in other places mentions DER-encoded specific ASN.1 structures and here it does not. The spec also specifies a type of bigint, but then in the reference example passes the value as a bytearray (as you would for a DER encoded value). In classic style it does not give a sample value.

I've read RFC5915. The first cut of this code used PyASN and had a lot more manual implementation than the current asn1crypto version. I think that's how I came to the conclusion on type here.

There is a unit test for the decode function, so it is loading a key into SoftHSM. However I notice the signing was FIXMEd out. I can't remember why this didn't work. It could have been an implementation bug inside SoftHSM of course.

**Comment by fzonneveld** [comment]
Maybe EC was not supported yet at the time of writing. I have just loaded the key into SoftHSMV2 and performed EDSA signature.  WIll need to copy and paste some sample code together. Without my modifications is does not work...

**Comment by fzonneveld** [comment]
First start SoftHSMV2 and initialize slot 0:
```
export SOFTHSM2_CONF=`realpath <config file>`
softhsm2-util --init-token --slot 0 --label "My token 1"
```
Where config file is normal config file, for me:
```
# SoftHSM v2 configuration file

directories.tokendir = /home/fzonneveld/tokens
objectstore.backend = file

# ERROR, WARNING, INFO, DEBUG
```
Pin will be asked for....
Make sure tokens dir exists in home dir.
```
     1  import os
     2  import uuid
     3  import asn1crypto.x509
     4  from asn1crypto.keys import ECPrivateKey
     5
     6  import click
     7
     8  import pkcs11
     9  from pkcs11.util import ec
    10  from pkcs11 import (
    11      Attribute, Mechanism, UserAlreadyLoggedIn
    12  )
    13  from cryptography.hazmat.primitives.asymmetric import utils
    14  from cryptography.hazmat.primitives.asymmetric.ec import \
    15      EllipticCurvePrivateKey, ECDSA
    16  from cryptography.hazmat.primitives.serialization import load_der_private_key
    17  from cryptography.hazmat.primitives import hashes
    18  from cryptography.hazmat.backends import default_backend
    19  from cryptography.exceptions import InvalidSignature
    20
    21
    22  def _get_session(module, slot, pin):
    23      lib = pkcs11.lib(module)
    24      token = lib.get_slots(token_present=True)[slot].get_token()
    25      try:
    26          session = token.open(rw=True, user_pin=pin)
    27      except UserAlreadyLoggedIn:
    28          session = token.open(rw=True)
    29      return session
    30
    31
    32  class ECSignature(asn1crypto.core.Sequence):
    33      _fields = [
    34          ('x', asn1crypto.core.Integer),
    35          ('y', asn1crypto.core.Integer)
    36      ]
    37
    38
    39  @click.command()
    40  @click.option('-m', '--module', default=os.environ.get('PKCS11_LIB'))
    41  @click.option('-s', '--slot', default=0)
    42  @click.option('-p', '--pin-file', type=click.File('r'),
    43                default=os.getenv('PIN_FILE'))
    44  @click.argument('private_key', type=click.File('rb'))
    45  def test_private_key(module, slot, pin_file, private_key):
    46      # Check the private key
    47      if pin_file is None:
    48          raise click.ClickException(
    49              'PIN file must be set with options or with'
    50              ' environment variable PIN_FILE')
    51      try:
    52          pk_der = private_key.read()
    53      except (TypeError, ValueError):
    54          raise click.ClickException('Invalid private key')
    55      session = _get_session(module, slot, pin_file.read().strip())
    56      label = uuid.uuid4().hex
    57
    58      pk_obj = load_der_private_key(
    59          pk_der, password=None, backend=default_backend())
    60      if isinstance(pk_obj, EllipticCurvePrivateKey):
    61          k = ec.decode_ec_private_key(pk_der)
    62          ec_pk = ECPrivateKey.load(pk_der)
    63          k[Attribute.EC_PARAMS] = ec_pk['parameters'].chosen.dump()
    64          k[Attribute.VALUE] = ec_pk['private_key'].contents
    65      else:
    66          raise click.ClickException('Invalid key type')
    67      k[Attribute.LABEL] = label
    68      k[Attribute.ID] = b'\x00'
    69      k[Attribute.TOKEN] = True
    70      k[Attribute.UNWRAP] = False
    71      k[Attribute.SIGN] = True
    72      k[Attribute.EXTRACTABLE] = False
    73      priv_key = session.create_object(k)
    74      click.echo(f"Inserted key with label={label}")
    75      h = hashes.SHA256()
    76      bin_data = os.urandom(h.digest_size)
    77      signature = priv_key.sign(bin_data, mechanism=Mechanism.ECDSA)
    78
    79      len_point = len(signature) // 2
    80      x = int.from_bytes(signature[:len_point], "big")
    81      y = int.from_bytes(signature[len_point:], "big")
    82
    83      sig = ECSignature({'x': x, 'y': y})
    84      s = sig.dump()
    85
    86      pub_key = pk_obj.public_key()
    87      click.echo('Validating signature')
    88
    89      try:
    90          pub_key.verify(
    91              s,
    92              bin_data,
    93              ECDSA(utils.Prehashed(hashes.SHA256()))
    94          )
    95      except InvalidSignature as e:
    96          print(str(e))
    97          raise click.ClickException('Signature failed....')
    98      click.echo('Verification OK')
    99      priv_key.destroy()
   100
   101
   102  if __name__ == '__main__':
   103      test_private_key()
```
In line 64 and 63 we resolve the key. When removed, will fail during import of key.
Can be run using (needs cryptography, click, asn1crypto and off course python-pkcs11) as follow:
```
 PKCS11_LIB=/usr/lib/x86_64-linux-gnu/softhsm/libsofthsm2.so python t.py ec.der --pin-file <(echo 1234)
```
Pin is 1234 and ec.der contains the key exported as described in previous comment.

**Comment by danni** [comment]
This is very compelling. I think you're right, that it wasn't correct in SoftHSM and so I disabled part of the test and forgot.

Do you want to see if you can replicate this in the test case `test_import_key_pair`? And if so send a PR for `test_import_key_pair` and `decode_ec_private_key`?

**Comment by fzonneveld** [comment]
I will create a PR for this, can close the comment.

**Comment by fzonneveld** [comment]
Reopened for PR.

**Comment by danni** [comment]
Fixed in #79
